networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/ResponseRpcDecoderTest.java

```diff
@@ -0,0 +1,195 @@
+    codec.onDataReceived(buffer(SUCCESS_CODE));
+    codec.onDataReceived(buffer(LENGTH_PREFIX));
+    // Split message data into a number of separate packets
+    for (int i = 0; i < MESSAGE_DATA.size(); i += 5) {
+      codec.onDataReceived(buffer(MESSAGE_DATA.slice(i, Math.min(MESSAGE_DATA.size() - i, 5))));
+    }
+    verifySingleMessageReceivedSuccessfully();
+  }
+
+  @Test
+  public void shouldParseMultipleResponsesFromSinglePacket() throws Exception {
+    final BeaconBlocksByRootRequestMessage secondMessage = createRequestMessage(2);
+    final Bytes secondMessageData = SimpleOffsetSerializer.serialize(secondMessage);
+    codec.onDataReceived(
+        buffer(
+            SUCCESS_CODE,
+            LENGTH_PREFIX,
+            MESSAGE_DATA,
+            SUCCESS_CODE,
+            getLengthPrefix(secondMessageData.size()),
+            secondMessageData));
+
+    verify(callback).accept(MESSAGE);
+    verify(callback).accept(secondMessage);
+
+    codec.close();
+    verifyNoMoreInteractions(callback);
+  }
+
+  @Test
+  public void shouldParseMultipleResponsesWhereSecondMessageIsSplit() throws Exception {
+    final BeaconBlocksByRootRequestMessage secondMessage = createRequestMessage(2);
+    final Bytes secondMessageData = SimpleOffsetSerializer.serialize(secondMessage);
+    codec.onDataReceived(buffer(SUCCESS_CODE, LENGTH_PREFIX, MESSAGE_DATA, SUCCESS_CODE));
+    verify(callback).accept(MESSAGE);
+
+    codec.onDataReceived(buffer(getLengthPrefix(secondMessageData.size()), secondMessageData));
+    verify(callback).accept(secondMessage);
+
+    codec.close();
+    verifyNoMoreInteractions(callback);
+  }
+
+  @Test
+  public void shouldThrowErrorIfMessagesHaveTrailingData() throws Exception {
+    codec.onDataReceived(
+        buffer(SUCCESS_CODE, LENGTH_PREFIX, MESSAGE_DATA, Bytes.fromHexString("0x1234")));
+    verify(callback).accept(MESSAGE);
+
+    assertThatThrownBy(codec::close).isEqualTo(RpcException.INCORRECT_LENGTH_ERROR);
+  }
+
+  @Test
+  public void shouldThrowErrorIfStatusCodeIsNotSuccess() throws Exception {
+    assertThatThrownBy(
+            () ->
+                codec.onDataReceived(
+                    buffer(ERROR_CODE, ERROR_MESSAGE_LENGTH_PREFIX, ERROR_MESSAGE_DATA)))
+        .isEqualTo(new RpcException(ERROR_CODE.get(0), ERROR_MESSAGE));
+    codec.close();
+  }
+
+  @Test
+  public void shouldNotThrowErrorFromCloseAfterErrorAlreadyThrown() throws Exception {
+    assertThatThrownBy(
+            () ->
+                codec.onDataReceived(
+                    buffer(
+                        ERROR_CODE,
+                        ERROR_MESSAGE_LENGTH_PREFIX,
+                        ERROR_MESSAGE_DATA,
+                        Bytes.fromHexString("0x1234"))))
+        .isInstanceOf(RpcException.class);
+
+    // Should not throw an exception despite the trailing data.
+    codec.close();
+  }
+
+  @Test
+  public void shouldReleaseByteBufsFromParsedMessages() throws Exception {
```

Comment:
Should we add similar tests that check ByteBuf refs to `RequestRpcDecoderTest`?

Reply:
The request decoder will only ever receive a single message so it only ever releases ByteBufs when it's done anyway.  The `MessageBuffer` class also has tests around this and all the tests confirm that the buffers are released by the end of the test (there's an `@After` in the shared parent class).